pkg/internalapis/csinodetopology/csinodetopology.go

```diff
@@ -16,9 +16,240 @@ limitations under the License.
+				log.Errorf("failed to get kubeconfig with error: %v", err)
+				return nil, err
+			}
+
+			// Create client to API server.
+			k8sclient, err := k8s.NewClientForGroup(ctx, config, csinodetopologyv1alpha1.GroupName)
+			if err != nil {
+				log.Errorf("failed to create k8sClient with error: %v", err)
+				return nil, err
+			}
+
+			// Create watcher for CSINodeTopology instances
+			crWatcher, err := k8s.NewCSINodeTopologyWatcher(ctx, config)
+			if err != nil {
+				log.Errorf("failed to create a watcher for CSINodeTopology CR. Error: %+v", err)
+				return nil, err
+			}
+
+			volumeTopologyInstance = &volumeTopology{
+				k8sClient:              k8sclient,
+				csiNodeTopologyWatcher: crWatcher,
+			}
+		}
+	} else {
+		volumeTopologyInstanceLock.RUnlock()
+	}
+
+	return volumeTopologyInstance, nil
+}
+
+// GetNodeTopologyLabels uses the CSINodeTopology CR to retrieve topology information of a node.
+func (volTopology *volumeTopology) GetNodeTopologyLabels(ctx context.Context, info *NodeInfo) (
+	map[string]string, error) {
+	log := logger.GetLogger(ctx)
+
+	// Fetch node object to set owner ref.
+	nodeObj, err := getNodeObject(ctx, info.NodeName)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	// Create spec for CSINodeTopology.
+	csiNodeTopologySpec := &csinodetopologyv1alpha1.CSINodeTopology{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: info.NodeName,
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: "v1",
+					Kind:       "Node",
+					Name:       nodeObj.Name,
+					UID:        nodeObj.UID,
+				},
+			},
+		},
+		Spec: csinodetopologyv1alpha1.CSINodeTopologySpec{
+			NodeID: info.NodeID,
+		},
+	}
+	// Create CSINodeTopology CR for the node.
+	err = volTopology.k8sClient.Create(ctx, csiNodeTopologySpec)
+	if err != nil {
+		if !apierrors.IsAlreadyExists(err) {
+			return nil, logger.LogNewErrorCodef(log, codes.Internal,
+				"failed to create CSINodeTopology CR. Error: %+v", err)
+		}
+	} else {
+		log.Infof("Successfully created a CSINodeTopology instance for NodeName: %q", info.NodeName)
+	}
+
+	timeoutSeconds := int64((time.Duration(getCSINodeTopologyWatchTimeoutInMin(ctx)) * time.Minute).Seconds())
+	watchCSINodeTopology, err := volTopology.csiNodeTopologyWatcher.Watch(metav1.ListOptions{
+		FieldSelector:  fields.OneTermEqualSelector("metadata.name", info.NodeName).String(),
+		TimeoutSeconds: &timeoutSeconds,
+		Watch:          true,
+	})
+	if err != nil {
+		return nil, logger.LogNewErrorCodef(log, codes.Internal,
+			"failed to watch on CSINodeTopology instance with name %q. Error: %+v", info.NodeName, err)
+	}
+	defer watchCSINodeTopology.Stop()
+
+	// Check if status gets updated in the instance within the given timeout seconds.
+	for event := range watchCSINodeTopology.ResultChan() {
+		csiNodeTopologyInstance, ok := event.Object.(*csinodetopologyv1alpha1.CSINodeTopology)
+		if !ok {
+			log.Warnf("Received unidentified object - %+v", event.Object)
+			continue
```

Comment:
I think this is triggered when the watch channel is closed, so we would need to return from here rather than continue. Have you tested that scenario with the recent changes?

Reply:
Yes tested this out with recent changes. The `Watch: true` parameter that I set on line 172 changes the behavior in which we receive events. Here, instead of a channel, I get a stream of events. You might be referring to the former way of calling a watch.